File: 0251. Flatten 2D Vector/Solution.cpp
#include <vector>
using namespace std;

class Vector2D {
public:
    Vector2D(vector<vector<int>>& vec)
        : it(vec.begin())
        , end_it(vec.end())
    {
    }

    int next()
    {
        if (i == (*it).size()) {
            it++;
            i = 0;
        }
        return (*it)[i++];
    }

    bool hasNext()
    {
        while (it != end_it && i == (*it).size()) {
            it++;
            i = 0;
        }
        return it != end_it;
    }

private:
    vector<vector<int>>::iterator it, end_it;
    size_t i = 0;
};

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D* obj = new Vector2D(vec);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: 0251. Flatten 2D Vector/Solution.py
from typing import List


class Vector2D:
    def __init__(self, vec: List[List[int]]):
        def generator():
            for row in vec:
                for ele in row:
                    yield ele

        self.gen = generator()
        self.next_val = next(self.gen, None)

    def next(self) -> int:
        res = self.next_val
        self.next_val = next(self.gen, None)
        return res

    def hasNext(self) -> bool:
        return self.next_val is not None


# Your Vector2D object will be instantiated and called as such:
# obj = Vector2D(vec)
# param_1 = obj.next()
# param_2 = obj.hasNext()
